sound/soc/sof/ipc4-topology.c

```diff
@@ -1658,6 +1703,13 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
 	if (ret < 0)
 		return ret;
 
+	if  (process->payload_with_output_fmt) {
```

Comment:
@libinyang IIUC this flag is only applicable for modules that are built into the base FW. Why do we only update the pipeline params only for base FW modules and not all external modules?

Reply:
@ranj063, I don't get  why we are locking module features to their position (in or out of basefw). When loaded to the DSP it is there, does not matter if it was loaded as part of the base library or as an external library.

This flag is telling that the given module's MOD_INIT_INSTANCE message should also have this extra output format section, it is irrelevant how that module got into the RAM of the DSP.

imho.

Reply:
@ujfalusi you are right once it is in the DSP, it is there. But this part is before it is initialized in the DSP and how we initialize them is very module implementation specific. Unfortunately for us, there is a difference between how a module in the base FW in initialized and how external module is initialized.

But to clarify my comment, my point was that updating the output params in the pipeline params should be independent of whether the module is built-in or not. And thie flags "payload_with_output_format) is only applicable to the built-in ones. So the check is restrictive and it shouldnt be.

Reply:
@ranj063, I don't think so. The library loading is a separate step and after that is done the module is in the DSP. Only modules which are already loaded can be initialized. loading a module to firmware and the module initialization is completely different thing, or it is supposed to be.

Reply:
I think @ranj063 is talking the IADK modules, which use different payload format with base FW module. I didn't support IADK modules in this PR, because they are using different payload format and SmartAMP and WoV doesn't need to update the params for IADK modules and don't have any scenarios to test the feature. We will add the IADK support when it is needed.

Reply:
@libinyang I think we should update the pipeline params based on the output format for all process modules unconditionally. It doesn't make sense to restrict it to only those that have the output format in their init module instance ipc payload.

Reply:
> @libinyang I think we should update the pipeline params based on the output format for all process modules unconditionally. It doesn't make sense to restrict it to only those that have the output format in their init module instance ipc payload.

Yes, this is our final goal, to support all process modules. However, we currently don't have any user case to test it. I think it should be easy to add such support. @aiChaoSONG had a patch before. The 2 types of modules has different implmentation to support this feature. @aiChaoSONG had a patch before. I think when we need to add this feature for IADK module, we can submit another PR for it based on this PR. What do you think?